### ISSUE TITLE
Research how to provide a better animations experience for Popover based components

### ISSUE THREAD

**Issue by cwoolum** [qwikifiers/qwik-ui#913]
Because we use the native popover API as the core of some of our components, there are limitations on how animations work. Standard animations placed on a popover based component won't work because the popover immediately disappears. There is specific documentation on how to use animations with popovers.

Work needs to be done to determine if this is the best approach or if we can find another workaround. The solution should also focus on providing a great DX so users don't struggle to implement.

**Comment by cwoolum** [comment]
After an analysis, here are my results

# Animations with the Popover API

## Overview


## 1. Popovers: Using `onBeforeToggle`

For popovers, the `onBeforeToggle` event provides a hook to apply animations before the popover opens or closes. This allows you to control the animation timeline, ensuring that the animation completes before any DOM changes occur.

### Example:

```javascript
export async function supportShowAnimation(popover: HTMLElement, isPolyfill: boolean) {
  popover.classList.add('popover-showing');
  popover.classList.remove('popover-closing');
  popover.dataset.open = '';
  popover.removeAttribute('data-closing');
  popover.removeAttribute('data-closed');
}

export function supportClosingAnimation(popover: HTMLElement) {
  popover.classList.remove('popover-showing');
  popover.classList.add('popover-closing');
  popover.dataset.closing = '';

  const { animationDuration, transitionDuration } = getComputedStyle(popover);

  const runAnimationEnd = () => {
    popover.classList.remove('popover-closing');
    popover.removeAttribute('data-closing');
    popover.dataset.closed = '';
  };

  const runTransitionEnd = () => {
    popover.classList.remove('popover-closing');
    popover.removeAttribute('data-closing');
    popover.dataset.closed = '';
  };

  if (animationDuration !== '0s') {
    popover.addEventListener('animationend', runAnimationEnd, { once: true });
  } else if (transitionDuration !== '0s') {
    popover.addEventListener('transitionend', runTransitionEnd, { once: true });
  } else {
    popover.classList.remove('popover-closing');
    popover.removeAttribute('data-closing');
    popover.dataset.closed = '';
  }
}
```
And the event handler
```jsx
onBeforeToggle$={[
        $(async (e) => {
          console.log('toggling')
          if (!context.panelRef?.value) return;

          if (e.newState === 'open' && context.panelRef.value) {
            await supportShowAnimation(context.panelRef.value, isPolyfillSig.value);
          }
          if (e.newState === 'closed') {
            supportClosingAnimation(context.panelRef.value);
          }
        }),
        props.onBeforeToggle$,
      ]}

```

This approach works the best but requires some very specific configurations to ensure everything works correctly. For transitions, the approach is also not bulletproof and may lead to an unpleasant UI in some cases

| Component | Event Name   | Event Type | Greenfield Browsers | Polyfilled Browsers | Comments                                    |
|-----------|--------------|------------|---------------------|---------------------|---------------------------------------------|
| Popover   | data-open    | Transition | ✅                   | ✅                   |                                             |
| Popover   | data-closing | Transition | ❌                   | ✅                   | In greenfield, transition flickers on close. Briefly re-opens and then immediately disappears again |
| Popover   | data-closed  | Transition | ✅                   | ❌                   | Doesn't ever close in polyfill browser      |
| Tooltip   | data-open    | Transition | ✅                   | ✅                   |                                             |
| Tooltip   | data-closing | Transition |  ❌                  | ✅                   | In greenfield, transition flickers on close. Briefly re-opens and then immediately disappears again |
| Tooltip   | data-closed| Transition | ✅                   | ❌                   | In polyfilled, transition never closes |
| Popover   | data-open    | Animation|      ✅              |             ✅       |                                             |
| Popover   | data-closing | Animation|       ✅             |          ✅          | |
| Popover   | data-closed  | Animation|        ❌            |         ❌           | Popover never closes in either   |
| Tooltip   | data-open    | Animation|      ✅              |           ✅        |                                             |
| Tooltip   | data-closing | Animation|     ✅               |         ✅           | |
| Tooltip   | data-closed| Animation|      ❌              |          ❌          |  Neither tooltip ever closes |


## 2. Applying Styles to :popover-open
The `:popover-open` pseudoclass (not to be confused with the `popover-open` data attribute is the native way of styling the popover but isn't supported in polyfilled browsers.

| Component |Event Name   |  Event Type | Greenfield Browsers | Polyfilled Browsers | Comments                                    |
|-----------|--------------|------------|---------------------|---------------------|---------------------------------------------|
| Popover   | open    | Transition | ✅                   | ❌                   |Popover never opens or closes with transition defined   |
| Popover   | closing | Transition | ✅                   | ❌                   | Popover never opens or closes with transition defined |
| Tooltip   | open    | Transition | ✅                   | ❌                   |     Popover never opens or closes with transition defined                                        |
| Tooltip   | closing | Transition | ✅                   | ❌                   |Popover never opens or closes with transition defined |
| Popover   | open    | Animation|      ✅              |             ❌                          |Popover opens in polyfilled but then immediately closes again. |
| Popover   | closing | Animation|       ✅             |          ❌                             |Popover opens in polyfilled but then immediately closes again. Greenfield requires special animation configuration to stay open on close. |
| Tooltip   | open    | Animation|      ✅              |           ❌        |          Popover opens in polyfilled but then immediately closes again.                                   |
| Tooltip   | closing | Animation|     ✅               |         ❌           | Popover opens in polyfilled but then immediately closes again.|

**Comment by maiieul** [comment]
So is there anything we can do and should do @cwoolum in your opinion?

**Comment by cwoolum** [comment]
Closed with #970
